### ISSUE TITLE
[Build]  C++ API calling fail: error C2280: 'Ort::Value::Value(const Ort::Value &)' : attempt to reference a deleted function

### ISSUE THREAD

**Issue by jinfagang** [microsoft/onnxruntime#15418]
### Describe the issue

error C2280: 'Ort::Value::Value(const Ort::Value &)' : attempt to reference a deleted function

### Urgency

Very urgent

### Target platform

windows

### Build script

noio

### Error / output

error C2280: 'Ort::Value::Value(const Ort::Value &)' : attempt to reference a deleted function

Here is the c++ caling code:

```c++
 std::vector<Ort::Value> ort_inputs;
  for (int i = 0; i < inputNames.size(); ++i) {
    auto ort_value = Ort::Value::CreateTensor<float>(
        memoryInfo, static_cast<float *>(inputs[i].data), inputs[i].get_size(),
        inputShapes[i].data(), inputShapes[i].size());
    assert(ort_value.IsTensor());
    ort_inputs.emplace_back(ort_value);
  }

try {
    std::vector<Ort::Value> ort_outputs;
    ort_outputs = session.Run(
        Ort::RunOptions{nullptr}, inputNames.data(), ort_inputs.data(),
        inputNames.size(), outputNames.data(), outputNames.size());
}
```

can not build

### Visual Studio Version

2019

**Comment by baijumeswani** [comment]
Would you please try to create the ort value in place in the vector?

```cc
std::vector<Ort::Value> ort_inputs;
  for (int i = 0; i < inputNames.size(); ++i) {
    ort_inputs.emplace_back(Ort::Value::CreateTensor<float>(
        memoryInfo, static_cast<float *>(inputs[i].data),
        inputs[i].get_size(), inputShapes[i].data(), inputShapes[i].size()));
  }

...
```

**Comment by jinfagang** [comment]
@baijumeswani  Hi,  I have tried, but not work, the problem caused is second part code:

```
try {
    std::vector<Ort::Value> ort_outputs;
    ort_outputs = session.Run(
        Ort::RunOptions{nullptr}, inputNames.data(), ort_inputs.data(),
        inputNames.size(), outputNames.data(), outputNames.size());
}
```

I don't know where caused the above exception.

**Comment by RyanUnderhill** [comment]
My guess is you're trying to copy the vector of outputs returned by Run() into ort_outputs.

Try this:

`std::vector<Ort::Value> ort_outputs = session.Run(...)`

**Comment by jinfagang** [comment]
@RyanUnderhill  hi, if I am not dosing so, how can I get the outputs?

**Comment by RyanUnderhill** [comment]
True, those actually should be identical and the return value should just be a move, so no copying is needed.

Can you see what line the error is happening on? And what version of C++ you're compiling with?

**Comment by jinfagang** [comment]
@RyanUnderhill  hello, I am using MSVC 2019 version, the error messages comes from onnxruntime header so I don't know eactly which line caused issue.

**Comment by jinfagang** [comment]
@RyanUnderhill unfortrunately MSVC didnt give me which line (my code) got above errors.

**Comment by iPengXPro** [comment]
I got the same problem,did you solve this problem?

**Comment by RyanUnderhill** [comment]
> 

In your case do you know what line the problem happened on?

**Comment by raisinglc03** [comment]
I got the same problem. Although I solved it, I have no idea why...

The following code return "error: use of deleted function ‘Ort::Value::Value(const Ort::Value&)"
{ ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }

'''
vector<Ort::Value> outputtensors;
auto outputtensor = Ort::Value::CreateTensor<float>( 
                memoryInfo, tmpTensorValue.data(), tmpOutTensorSize, tmpOutDim.data(), tmpOutDim.size()
            );
outputtensors.push_back(outputtensor);
'''

I put CreateTensor inside push_back and then it work ...

'''
vector<Ort::Value> outputtensors;
outputtensors.push_back( Ort::Value::CreateTensor<float>( 
                memoryInfo, tmpTensorValue.data(), tmpOutTensorSize, tmpOutDim.data(), tmpOutDim.size()
            );

'''

**Comment by RyanUnderhill** [comment]
@raisinglc03 

Ah, I see what the problem would be in your case. push_back is trying to make a copy of 'outputtensor', if you push_back with a std::move() it should work:

```
vector<Ort::Value> outputtensors;
auto outputtensor = Ort::Value::CreateTensor<float>( 
                memoryInfo, tmpTensorValue.data(), tmpOutTensorSize, tmpOutDim.data(), tmpOutDim.size()
            );
outputtensors.push_back(std::move(outputtensor));
```

**Comment by raisinglc03** [comment]
> @raisinglc03
> 
> Ah, I see what the problem would be in your case. push_back is trying to make a copy of 'outputtensor', if you push_back with a std::move() it should work:
> 
> ```
> vector<Ort::Value> outputtensors;
> auto outputtensor = Ort::Value::CreateTensor<float>( 
>                 memoryInfo, tmpTensorValue.data(), tmpOutTensorSize, tmpOutDim.data(), tmpOutDim.size()
>             );
> outputtensors.push_back(std::move(outputtensor));
> ```

Thanks! It work!

**Comment by clementruhm** [comment]
I want to keep the output tensors for future use. Just declaring the vector<Ort::Value> in the header, works in Linux/Mac, but for Windows (visual studio 2022) fails with the error above. Any suggestion on how you can keep the output tensors as member of a class?

**Comment by clementruhm** [comment]
for anyone struggling with similar, it occurred I had to delete copy constructor and assignment operator for the class that holds Ort::Value vector:

```
    ONNXRequest(const ONNXRequest& other) = delete;
    ONNXRequest& operator=(const ONNXRequest& other) = delete;
```
